assets/test/e2e/test.js

```diff
@@ -33,9 +33,100 @@ describe('', function() {
+    browser.wait(protractor.ExpectedConditions.elementToBeClickable(nextButton), 2000);
+    return nextButton.click();
+  }
+
+  var requestCreateFromTemplate = function(projectName, templateName) {
+    var uri = '/project/' + projectName + '/create';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    var template = element(by.cssContainingText('.catalog h3 > a', templateName));
+    expect(template.isPresent()).toBe(true);
+    return template.click();
+  }
+
+  var attachBuilderImageToSource = function(projectName, builderImageName) {
+    var uri = '/project/' + projectName + '/catalog/images';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.cssContainingText('h1', "Select a builder image")).isPresent()).toBe(true);
+    var builderImageLink = element(by.cssContainingText('div > h3 > a', builderImageName));
+    expect(builderImageLink.isPresent()).toBe(true);
+    return builderImageLink.click();
+  }
+
+  var createFromSourceSummary = function(projectName, builderImageName, appName) {
+    var uri = '/project/' + projectName + '/create/fromimage';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.css('.create-from-image h1')).getText()).toEqual(builderImageName);
+    expect(element(by.cssContainingText('div > h2', "Name")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Routing")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Deployment Configuration")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Build Configuration")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Scaling")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Labels")).isPresent()).toBe(true);
+    element(by.name('appname')).clear().sendKeys(appName);
+    return element(by.buttonText("Create")).click();
+  }
+
+  var createFromTemplateSummary = function(projectName, templateName, parameterNames, labelNames) {
+    var uri = '/project/' + projectName + '/create/fromtemplate';
+    expect(browser.getCurrentUrl()).toContain(uri);
+    expect(element(by.css('.create-from-template h1')).getText()).toEqual(templateName);
+    expect(element(by.cssContainingText('div > h2', "Images")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Parameters")).isPresent()).toBe(true);
+    expect(element(by.cssContainingText('div > h2', "Labels")).isPresent()).toBe(true);
+    if (parameterNames) {
+      for (i = 0; i < parameterNames.length; i++) {
+        expect(element(by.cssContainingText('.env-variable-list label.key', parameterNames[i])).isPresent()).toBe(true);
+      }
+    }
+    if (labelNames) {
+      for (i = 0; i < labelNames.length; i++) {
+        expect(element(by.cssContainingText('.label-list span.key', labelNames[i])).isPresent()).toBe(true);
+      }
+    }
+    return element(by.buttonText("Create")).click();
```

Comment:
these returning what looks like promises is a little weird... if one of the above conditions fails, does control reach this point or is an exception thrown?

I thought a lot of the tests already had timeouts/retries built into them... I'm not sure what combining that with promises gets us